Simplify cache flushing

jsumners opened the pull request:
This commit eliminates checking the stream against `process.stdout`
and `process.stderr` when flushing the cache on exit. It does so
by forcing the stream into blocking mode by setting the stream's
internal `_handle` to blocking mode via the V8 `setBlocking` method.

I opted to unblock the stream after writing out the cache just to be
comprehensive. I don't believe it is necessary, but I'd rather not
be the cause of the application hanging indefinitely due to some
edge case as a result of the stream being in blocking mode.

jsumners commented:
Hmm, hold off on this. I forgot that `stream` may be supplied by the user and is not necessarily a stdio stream. I need to determine if this is a valid solution or should be expanded.

jsumners commented:
After further investigation I have come to the conclusion that this change is not worth it. The current implementation is less complex that would be necessary to verify that the output stream is capable of being set to blocking mode.

mcollina commented:
@jsumners I think we still have a bug in the current implementation: using extreme mode without anything that is not a `tty` will result in the output not being flushed correctly. A contribution in that regard (even with a `throw` in the constructor, plus a README entry) would be wonderful.

mcollina commented:
Tests are failing on v0.10 and v0.12: https://travis-ci.org/mcollina/pino/jobs/158851528#L263-L277.
If it's simple, we can keep the support going (it seems it's just a `const`).

Regarding the semantic version range of this, it might be semver-major. If that's the case, we can ditch v0.10 support there as well. We can also consider this semver-minor, as whoever was using `extreme: true` without a `tty` would have been _extremely disappointed_ by a data loss.

What do you think?

jsumners commented:
Oops. I'm a `let`/`const` guy; I usually do a good job at following a project's code style but that one slipped in without me realizing it. I amended the last commit to remove it.

I think the case can be made for either versioning on this. `minor` due to the "bug" nature but `major` due to the radical change. I suppose my answer depends on your opinion of dropping `var` altogether. I'm willing to update the whole code base to remove the usage of `var` and replace it with `let` and `const` as appropriate. Is that a style you guys like? If so, it would drop support for 0.12 as well.

Personally, I think dropping support for 0.10 *and* 0.12 is long overdue, regardless of the code style.

mcollina commented:
`let` is slow on v4, so `let` is v6+ only from this library. I can get in a PR for const for v3 (whenever that happens, at the latest when v0.12 goes out of support).

I think we should semver-minor it, as it is an edge case anyway.

mcollina commented on pino.js:
```diff
@@ -65,6 +65,10 @@ function pino (opts, stream) {
 
+  if (cache && stream.isTTY === false) {
+    throw new Error('extreme mode stream must be a tty stream')
+  }
```
This is not tested. It also does not work, because `isTTY` will be `undefined`.

mcollina commented:
There is also the issue with the benchmarks. With the proposed changes, they won't work.

jsumners commented on test/extreme.test.js:
```diff
@@ -4,6 +4,7 @@ var test = require('tap').test
 var os = require('os')
 var writeStream = require('flush-write-stream')
 var spawn = require('child_process').spawn
```
The scripts passed into `spawn`, even with `{stdio: [0, 1, 2]}`, are not receiving the actual process streams. Thus, the check for a tty is failing (after having been corrected to check properly). Do you have any suggestions for spawning the test script with a tty? Forking results in the same issue.

mcollina replied:
Another option is to get the file descriptor, and do a `writeSync` https://github.com/nodejs/node/blob/master/lib/internal/fs.js#L50.

I think it might be a better approach rather than `setBlocking`  and require a `tty`.

jsumners replied:
The only way I'm seeing that will work is to check if the passed in stream has an `fd` property or `stream._handle.fd` property. Do you agree?

mcollina replied:
I agree. That's a lower assumption than a tty.

jsumners replied:
Actually, that still wouldn't work for the spawned process streams:

```
WriteStream {
  _writableState:
   WritableState {
 objectMode: false,
 highWaterMark: 16384,
 needDrain: false,
 ending: false,
 ended: false,
 finished: false,
 decodeStrings: true,
 defaultEncoding: 'utf8',
 length: 0,
 writing: false,
 corked: 0,
 sync: true,
 bufferProcessing: false,
 onwrite: [Function],
 writecb: null,
 writelen: 0,
 bufferedRequest: null,
 lastBufferedRequest: null,
 pendingcb: 0,
 prefinished: false,
 errorEmitted: false,
 bufferedRequestCount: 0,
 corkedRequestsFree: CorkedRequest { next: [Object], entry: null, finish: [Function] } },
  writable: true,
  domain: null,
  _events: {},
  _eventsCount: 0,
  _maxListeners: undefined,
  destroyed: false,
  _worker: [Function],
  _flush: null }
```

mcollina replied:
what about  `fork`?

mcollina commented on pino.js:
```diff
@@ -73,13 +83,9 @@ function pino (opts, stream) {
-        }
+        stream._handle.setBlocking(true)
+        stream.write(cache.buf)
+        stream._handle.setBlocking(false)
```
we should use the `fd` here, if `_handle` is not there.

jsumners commented:
I'm correcting some things in my last push. It'll show up here rebased when I believe it to be ready.

mcollina commented on README.md:
```diff
@@ -574,6 +574,8 @@ This has a couple of important caveats:
   * A sigkill (or other untrappable signal) will probably result in the same
   * If writing to a stream other than `process.stdout` or `process.stderr`, there is a slight possibility of lost logs or even partially written logs if the OS buffers don't have enough space, or something else is being written to the stream (or maybe some other reason we've not thought of)
+* If you supply an alternate stream to the constructor, then that stream must be a tty stream so that it can be properly
+flushed on exit
```
there is probably a space issue here

mcollina commented:
There was a bug in the code, for some reason `writeSync` assumes it's a `Buffer` and causes an abort. The bug was in `pino.js`, this works:

```js
require('fs').writeSync(fd, cache.buf)
```

You should also add some lines of comment before `writeSync` to explain _why_.

jsumners commented:
That was an embarrassing bug. The issue is corrected. 

I also discovered that `stream.fd` could be `null` under some circumstances. Specifically `fs.createWriteStream('/dev/null')` will result in a `null` file descriptor. So I updated the blockable check to make sure the file descriptor is a truthy value.

In turn, I updated all of the benchmarks to use a stream with a blocking file descriptor for the extreme benchmarks.

mcollina commented on benchmarks/basic.js:
```diff
@@ -9,8 +9,15 @@ var fs = require('fs')
+}
 delete require.cache[require.resolve('../')]
-var plogExtreme = require('../')({extreme: true}, dest)
+var plogExtreme = require('../')({extreme: true}, extremeDest)
```
this is not needed, we can write directly to a file.

jsumners replied:
Sure, we could do the following to get a stream with a direct file descriptor:

```js
var fd = fs.openSync('/tmp/extreme', 'w')
var extremeDest = fs.createWriteStream('/tmp/extreme', {fd: fd})
```

But just `fs.createWriteStream('/tmp/extreme')` (or to `/dev/null`) results in a stream without a direct file descriptor attached; the property is set to `null`. Thus we have to do *something* to get a blockable stream. I just opted for the TTY.

mcollina commented on benchmarks/childChild.js:
```diff
@@ -7,7 +7,11 @@ var fs = require('fs')
+var tty = require('tty')
+var extremeDest = new tty.WriteStream(1)
+extremeDest.write = function (s) { dest.write(s) }
+var plogExtreme = require('../')({extreme: true}, extremeDest).child({ a: 'property' }).child({sub: 'child'})
```
same here

mcollina commented on benchmarks/childCreation.js:
```diff
@@ -8,7 +8,11 @@ var fs = require('fs')
+var tty = require('tty')
+var extremeDest = new tty.WriteStream(1)
+extremeDest.write = function (s) { dest.write(s) }
+var plogExtreme = require('../')({extreme: true}, extremeDest)
```
and here

mcollina commented on benchmarks/deep-object.js:
```diff
@@ -8,12 +8,16 @@ var winston = require('winston')
 var plogUnsafe = require('../')({safe: false}, dest)
 delete require.cache[require.resolve('../')]
-var plogUnsafeExtreme = require('../')({extreme: true, safe: false}, dest)
+var plogUnsafeExtreme = require('../')({extreme: true, safe: false}, extremeDest)
```
:)

mcollina commented on benchmarks/multiArg.js:
```diff
@@ -9,12 +9,16 @@ var fs = require('fs')
 var plogUnsafe = require('../')({safe: false}, dest)
 delete require.cache[require.resolve('../')]
-var plogUnsafeExtreme = require('../')({extreme: true, safe: false}, dest)
+var plogUnsafeExtreme = require('../')({extreme: true, safe: false}, extremeDest)
```
now we support files

mcollina commented on benchmarks/object.js:
```diff
@@ -9,12 +9,16 @@ var fs = require('fs')
+
+var tty = require('tty')
+var extremeDest = new tty.WriteStream(1)
+extremeDest.write = function (s) { dest.write(s) }
```
:)

mcollina commented:
> In turn, I updated all of the benchmarks to use a stream with a blocking file descriptor for the extreme benchmarks.

I probably have put some comments in for those. Why?

jsumners commented:
You did. I responded to the first comment (this PR is getting crazy to follow). I'll reproduce it here for clarity:

Sure, we could do the following to get a stream with a direct file descriptor:

```js
var fd = fs.openSync('/tmp/extreme', 'w')
var extremeDest = fs.createWriteStream('/tmp/extreme', {fd: fd})
```

But just `fs.createWriteStream('/tmp/extreme')` (or to `/dev/null`) results in a stream without a direct file descriptor attached; the property is set to `null`. Thus we have to do *something* to get a blockable stream. I just opted for the TTY.

mcollina commented:
The file descriptor is populated asynchronously. Does it make sense to postpone that check for 1 second and then throw, _if it is `null`_?

Normally it would be `undefined` if the property is not present.

jsumners commented:
I don't think the logger should be postponing anything. That would make it cumbersome to instantiate a logger instance in all cases, not just for extreme mode. I think it is better to require valid stream be supplied than to accept any stream and play a guessing game as to whether or not it will eventually be viable.

mcollina commented:
I would really like to maintain the flow:

```js
var fs = require('fs')
var logger = pino({ extreme: true}, fs.createWriteStream('aaa'))
```

That's why I would like to _postpone_ the check.
Consider the following:

```js
var fs = require('fs')

var s = fs.createWriteStream('out')

var w = require('stream').Writable()

console.log('fd', s.fd, w.fd)

setTimeout(function () {
  console.log('fd', s.fd)
  console.log('fd', s.fd, w.fd)
}, 1000)
```

It outputs:

```
$ node b.js
fd null undefined
fd 10
fd 10 undefined
```

The `fd` property moves from `null` to a number, and then it goes to `undefined`.

jsumners commented:
Okay, but how do you propose alerting the user during construction when the check fails? The check itself is asynchronous, so we'd have to issue an event or invoke a callback. That will break the flow as well.

mcollina commented:
Oh, I don't think we should do a callback. We can just emit `'error'` on the stream. That will do fine in letting the user know.

jsumners commented:
So something similar to?:

```js
function pino (opts, stream) {
  setImmediate(function () {
    if (!stream.fd) stream.emit('error', 'unusable')
  })
}

var log = pino(astream)
astream.on('error', function (err) { process.exit(1) /* very sad */ })
```

mcollina commented:
emit a full `Error`, do the check only if `extreme: true` and add a quick note on the docs.

jsumners commented:
Of course. Just wanted to verify we're on the same page prior to implementing the change.

What do you think of emitting a different event? Maybe a `pino-error` event? That would allow users to know it is specifically pino bombing and not the actual stream. If not a special event, shall I prefix the error message with 'pino: '?

mcollina commented:
pino is an `EventEmitter` if we are in Node.js, just emit `'error'` on the `pino` instance. It's more coherent than doing it on the stream.

On the browser... `extreme: true` mode should probably be disabled entirely, don't bother for now, we'll do it on a later PR.

jsumners commented on README.md:
```diff
@@ -574,6 +574,18 @@ This has a couple of important caveats:
+    // do something with chunk
+  }}))
+  logger.on('error', function (err) {
+    // this callback _will_ be invoked for the type of stream we supplied
```
The comment is simply saying that due to the `stream.Writable` not exposing a file descriptor the error callback will be invoked. And `_will_` is underlined with a traditional ASCII underline for emphasis. Would you rather see this callback?:

```js
function (err) {
  console.error('pino logger cannot flush on exit due to provided output stream')
  process.exit(1)
}
```

mcollina replied:
Yes :).

jsumners replied:
Updated the commit with the change.

mcollina commented:
this is good to go. Very good job @jsumners!
I'll merge and release when I get a chance!

mcollina commented:
(wrong button)

jsumners commented:
I create issue #88 for proper tracking of the underlying bug.

mcollina commented:
Merged!
